File: lib/store.js
import { existsSync } from "node:fs";
import { appendFile, mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { CONFIG_FILE, DEFAULT_SETTINGS, GROUPS, SCHEMA_URL, SETTINGS_DIR, SETTINGS_FILE } from "./constants.js";

async function readJson(filePath, fallback) {
  if (!existsSync(filePath)) return fallback;
  return JSON.parse(await readFile(filePath, "utf8"));
}

export async function readConfig() {
  const config = await readJson(path.join(process.cwd(), CONFIG_FILE), {
    "$schema": SCHEMA_URL,
    groups: []
  });
  return normalizeConfig({ "$schema": config.$schema ?? SCHEMA_URL, ...config });
}

export async function writeConfig(config) {
  await writeFile(path.join(process.cwd(), CONFIG_FILE), `${JSON.stringify(serializeConfig(config), null, 2)}\n`);
}

export async function readSettings() {
  return { ...DEFAULT_SETTINGS, ...(await readJson(path.join(process.cwd(), SETTINGS_FILE), {})) };
}

export async function writeSettings(settings) {
  await mkdir(path.join(process.cwd(), SETTINGS_DIR), { recursive: true });
  await writeFile(path.join(process.cwd(), SETTINGS_FILE), `${JSON.stringify(settings, null, 2)}\n`);
  await ensureSettingsIgnored();
}

export async function ensureSettingsIgnored(cwd = process.cwd()) {
  const gitignorePath = path.join(cwd, ".gitignore");
  const entry = `${SETTINGS_DIR}/`;
  const current = existsSync(gitignorePath) ? await readFile(gitignorePath, "utf8") : "";
  if (current.split(/\r?\n/).includes(entry)) return;
  const prefix = current.length > 0 && !current.endsWith("\n") ? "\n" : "";
  await appendFile(gitignorePath, `${prefix}${entry}\n`);
}

export function commandId(command) {
  return command.id ?? `${command.groupId ?? command.group ?? "manual"}.${command.name}`;
}

export function normalizeGroup(group) {
  if (!group || typeof group.id !== "string") throw new Error(`Each group in ${CONFIG_FILE} needs an "id" string.`);
  const defaults = GROUPS[group.id] ?? {};
  return {
    id: group.id,
    label: group.label ?? defaults.label ?? group.id,
    description: group.description ?? defaults.description ?? "",
    icon: group.icon ?? defaults.icon ?? "",
    type: group.type ?? defaults.type ?? "manual",
    source: group.source ?? defaults.source,
    commands: (group.commands ?? []).map((command) => normalizeCommand(command, group.id))
  };
}

export function normalizeCommand(command, groupId = command?.groupId ?? command?.group ?? "manual") {
  if (!command || typeof command.name !== "string" || typeof command.command !== "string") {
    throw new Error(`Each command in ${CONFIG_FILE} needs "name" and "command" strings.`);
  }
  return {
    id: command.id ?? `${groupId}.${command.name}`,
    name: command.name,
    command: command.command,
    description: command.description ?? "",
    cwd: command.cwd ?? ".",
    groupId,
    group: groupId,
    source: command.source ?? GROUPS[groupId]?.source ?? "manual",
    imported: Boolean(command.imported)
  };
}

export function normalizeConfig(config) {
  if (Array.isArray(config.groups)) {
    return { "$schema": config.$schema ?? SCHEMA_URL, groups: config.groups.map(normalizeGroup) };
  }
  if (Array.isArray(config.commands)) {
    const groups = new Map();
    for (const command of config.commands.map((item) => normalizeCommand(item))) {
      const groupId = command.groupId;
      if (!groups.has(groupId)) groups.set(groupId, groupShell(groupId, command.imported));
      groups.get(groupId).commands.push(command);
    }
    return { "$schema": config.$schema ?? SCHEMA_URL, groups: [...groups.values()].map(normalizeGroup) };
  }
  throw new Error(`${CONFIG_FILE} must contain a "groups" array.`);
}

export function groupShell(groupId, imported = false) {
  const defaults = GROUPS[groupId] ?? {};
  return {
    id: groupId,
    label: defaults.label ?? groupId,
    description: defaults.description ?? "",
    icon: defaults.icon ?? "",
    type: defaults.type ?? (imported ? "imported" : "manual"),
    source: defaults.source,
    commands: []
  };
}

export function getGroups(config) {
  return normalizeConfig(config).groups;
}

export function getCommands(config) {
  return getGroups(config).flatMap((group) => group.commands.map((command) => normalizeCommand(command, group.id)));
}

export function serializeCommand(command) {
  const groupId = command.groupId ?? command.group ?? "manual";
  return {
    id: command.id ?? `${groupId}.${command.name}`,
    name: command.name,
    command: command.command,
    description: command.description ?? "",
    cwd: command.cwd ?? ".",
    source: command.source ?? GROUPS[groupId]?.source ?? "manual",
    imported: Boolean(command.imported)
  };
}

export function serializeGroup(group) {
  const normalized = normalizeGroup(group);
  return {
    id: normalized.id,
    label: normalized.label,
    description: normalized.description,
    icon: normalized.icon,
    type: normalized.type,
    ...(normalized.source ? { source: normalized.source } : {}),
    commands: normalized.commands.map(serializeCommand)
  };
}

export function serializeConfig(config) {
  return {
    "$schema": config.$schema ?? SCHEMA_URL,
    groups: getGroups(config).filter((group) => group.commands.length > 0).map(serializeGroup)
  };
}

export function replaceGroups(config, groups) {
  return serializeConfig({ ...config, groups });
}

export function findCommand(commands, identifier) {
  return commands.find((command) => {
    return commandId(command) === identifier || command.name === identifier || `${command.group}.${command.name}` === identifier;
  });
}

File: internal/config/store.go
package config

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

func Read(root string) (Config, error) {
	path := filepath.Join(root, ConfigFile)
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return Normalize(Config{Schema: SchemaURL, Groups: []Group{}})
	}
	if err != nil {
		return Config{}, fmt.Errorf("read %s: %w", path, err)
	}
	var value Config
	if err := json.Unmarshal(data, &value); err != nil {
		return Config{}, fmt.Errorf("parse %s: %w", path, err)
	}
	return Normalize(value)
}

func Serialize(input Config) (Config, error) {
	normalized, err := Normalize(input)
	if err != nil {
		return Config{}, err
	}
	result := Config{Schema: normalized.Schema, Groups: []Group{}}
	for _, group := range normalized.Groups {
		if len(group.Commands) == 0 {
			continue
		}
		serialized := group
		serialized.Commands = make([]Command, len(group.Commands))
		for i, command := range group.Commands {
			command.GroupID = ""
			command.Group = ""
			serialized.Commands[i] = command
		}
		result.Groups = append(result.Groups, serialized)
	}
	return result, nil
}

func Write(root string, input Config) error {
	value, err := Serialize(input)
	if err != nil {
		return err
	}
	data, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return fmt.Errorf("encode %s: %w", ConfigFile, err)
	}
	data = append(data, '\n')
	return os.WriteFile(filepath.Join(root, ConfigFile), data, 0o644)
}

func ReplaceGroups(input Config, groups []Group) (Config, error) {
	input.Groups = groups
	input.Commands = nil
	return Serialize(input)
}

func ReadSettings(root string) (Settings, error) {
	settings := DefaultSettings()
	path := filepath.Join(root, SettingsFile)
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return settings, nil
	}
	if err != nil {
		return Settings{}, fmt.Errorf("read %s: %w", path, err)
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return Settings{}, fmt.Errorf("parse %s: %w", path, err)
	}
	encoded, err := json.Marshal(settings)
	if err != nil {
		return Settings{}, err
	}
	var merged map[string]json.RawMessage
	if err := json.Unmarshal(encoded, &merged); err != nil {
		return Settings{}, err
	}
	for key, value := range raw {
		merged[key] = value
	}
	encoded, err = json.Marshal(merged)
	if err != nil {
		return Settings{}, err
	}
	if err := json.Unmarshal(encoded, &settings); err != nil {
		return Settings{}, fmt.Errorf("parse %s: %w", path, err)
	}
	return settings, nil
}

func WriteSettings(root string, settings Settings) error {
	dir := filepath.Join(root, SettingsDir)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("create %s: %w", dir, err)
	}
	data, err := json.MarshalIndent(settings, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(filepath.Join(root, SettingsFile), append(data, '\n'), 0o644); err != nil {
		return err
	}
	return EnsureSettingsIgnored(root)
}

func EnsureSettingsIgnored(root string) error {
	path := filepath.Join(root, ".gitignore")
	current, err := os.ReadFile(path)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return err
	}
	entry := []byte(SettingsDir + "/")
	for _, line := range bytes.Split(current, []byte{'\n'}) {
		if bytes.Equal(bytes.TrimSuffix(line, []byte{'\r'}), entry) {
			return nil
		}
	}
	if len(current) > 0 && current[len(current)-1] != '\n' {
		current = append(current, '\n')
	}
	current = append(current, entry...)
	current = append(current, '\n')
	return os.WriteFile(path, current, 0o644)
}
